File: rust/src/solutions/day04.rs
// What did I learn?
// fun with iterators, how to extract from them and how to check splits don't have extra values.

use std::str::FromStr;
use log::info;
use crate::Solver;

pub(crate) struct Solution {
    full_overlaps: u32,
    partial_overlaps: u32,
}

impl Solution {
    pub(crate) fn new() -> Solution {
        Solution {
            full_overlaps: 0,
            partial_overlaps: 0
        }
    }
}

struct Elf {
    min: u32,
    max: u32,
}

impl Elf {
    fn parse(sections: &str) -> Elf {
        let mut parts = sections.split("-")
            .map(|text| u32::from_str(text).unwrap());
        let elf = Elf { min: parts.next().unwrap(), max: parts.next().unwrap() };
        if parts.next().is_some() {
            panic!("More parts than expected {sections}");
        }
        elf
    }

    fn inside(&self, other: &Elf) -> bool {
        (other.min <= self.min && self.min <= other.max) || (other.min <= self.max && self.max <= other.max)
    }
}

impl Solver for Solution {
    fn parse(&mut self, line: &str) {
        let mut parts = line.split(",");
        let elf1 = Elf::parse(parts.next().unwrap());
        let elf2 = Elf::parse(parts.next().unwrap());
        if parts.next().is_some() {
            panic!("More parts than expected {line}");
        }
        // full overlap when both ends of one of the ranges are fully inside the other
        if (elf1.min <= elf2.min && elf1.max >= elf2.max) || (elf1.min >= elf2.min && elf1.max <= elf2.max) {
            self.full_overlaps += 1;
        }
        // partial overlap if either end of each range is inside the other (one check is enough?)
        if elf1.inside(&elf2) || elf2.inside(&elf1) {
            self.partial_overlaps += 1
        }
    }

    fn solve(&mut self) -> Option<(String, String)> {
        info!("[1] Fully overlapping sections: {}", self.full_overlaps);
        info!("[2] Partially overlapping sections: {}", self.partial_overlaps);
        Some((self.full_overlaps.to_string(), self.partial_overlaps.to_string()))
    }
}

File: python/day04.py
import logging

from advent import Solver


# https://adventofcode.com/2022/day/4
# overlap comparisons were probably cleaner if they were placed into functions
# the partial overlap was probably working instead calculated as inverted (= one segment ends being outside range of
# the other segment)

log = logging.getLogger("day.04")


def inside(pos, range):
    return range[0] <= pos <= range[1]


class Solution(Solver):
    def __init__(self):
        self.full_overlaps = 0
        self.partial_overlaps = 0

    def parse(self, line: str):
        pairs = line.split(",")
        assert len(pairs) == 2
        elf1 = [int(x) for x in pairs[0].split("-")]
        assert len(elf1) == 2
        elf2 = [int(x) for x in pairs[1].split("-")]
        assert len(elf2) == 2
        # full overlap when both ends of one of the ranges are fully inside the other
        if (elf1[0] <= elf2[0] and elf1[1] >= elf2[1]) or (elf1[0] >= elf2[0] and elf1[1] <= elf2[1]):
            self.full_overlaps += 1
        # partial overlap if either end of each range is inside the other (one check is enough?)
        if inside(elf1[0], elf2) or inside(elf1[1], elf2) or inside(elf2[0], elf1) or inside(elf2[1], elf1):
            self.partial_overlaps += 1

    def solve(self):
        log.info(f"Fully overlapping sections: {self.full_overlaps}")
        log.info(f"Partially overlapping sections: {self.partial_overlaps}")
        return str(self.full_overlaps), str(self.partial_overlaps)

